src/crystal/system/win32/file.cr

```diff
@@ -26,11 +26,77 @@ module Crystal::System::File
   end
 
   def self.open(filename : String, flags : Int32, perm : ::File::Permissions) : {LibC::Int, Errno}
-    flags |= LibC::O_BINARY | LibC::O_NOINHERIT
+    access, disposition, attributes, share = self.posix_to_open_opts flags, perm
 
-    fd = LibC._wopen(System.to_wstr(filename), flags, perm)
+    handle = LibC.CreateFileW(
+      System.to_wstr(filename),
+      access,
+      share, # UNIX semantics
+      nil,
+      disposition,
+      attributes,
+      LibC::HANDLE.null
+    )
+
+    if handle == LibC::INVALID_HANDLE_VALUE
+      return {-1, Errno.value}
+    end
+
+    fd = LibC._open_osfhandle handle, LibC::O_CREAT | LibC::O_TRUNC | LibC::O_RDWR | LibC::O_BINARY
+
+    if fd == -1
+      return {-1, Errno.value}
+    end
+
+    LibC._setmode fd, LibC::O_BINARY
+
+    {fd, Errno::NONE}
+  end
+
+  private def self.posix_to_open_opts(flags : Int32, perm : ::File::Permissions)
+    access = disposition = attributes = 0
+
+    case flags & (LibC::O_RDONLY | LibC::O_WRONLY | LibC::O_RDWR)
+    when LibC::O_RDONLY then access = LibC::GENERIC_READ
+    when LibC::O_WRONLY then access = LibC::GENERIC_WRITE
+    when LibC::O_RDWR   then access = LibC::GENERIC_READ | LibC::GENERIC_WRITE
+    end
```

Comment:
Is there any benefit from this structure over `flags & LibC::O_RDONLY != 0` etc. conditions which would IMO be more straightforward?

Also I suppose the assignment could be moved outside the conditional, with the `else` branch evaluating to `0`.
Totally optional, but I think these two changes would improve readability.